File: machine-learning-ex2/ex2/python/ex2.py
## Machine Learning Online Class (Python implementation) - Exercise 2: Logistic Regression
import matplotlib.pyplot as plt
import numpy as np
import scipy.optimize as op
from ex2lib import *

if __name__ == '__main__':
    data = np.loadtxt('ex2data1.txt', delimiter = ',')
    X = data[:,0:2]
    y = data[:,2]
## ==================== Part 1: Plotting ====================
    print("Plotting data with + indicating (y=1) examples, and o indicating (y=0) examples.")

    plotData(X, y)
    plt.xlabel('Exam 1 score')
    plt.ylabel('Exam 2 score')
    plt.legend(['Admitted', 'Not admitted'])
    plt.show()

    input("Program paused. Press enter to continue.")

## ============ Part 2: Compute Cost and Gradient ============
    # Setup the data matrix appropriately, and add ones for the intercept
    [m, n] = np.shape(X)

    # Add intercept term to x and X_test
    X = np.concatenate([np.ones([m, 1]), X], axis = 1)

    # Initialize fitting parameters
    initial_theta = np.zeros([n + 1, 1])

    # Compute and display initial cost and gradient
    [cost, grad] = costFunction(initial_theta, X, y)

    print("Cost at initial theta (zeros): %.3f" % cost)
    print("Gradient at initial theta (zeros):\n\t[%.3f, %.3f, %.3f]" %
            (grad[0], grad[1], grad[2]))

    input("Program paused. Press enter to continue.")

## ============= Part 3: Optimizing using fminunc  =============
    # Run op.minimize() to obtain the optimal theta
    res = op.minimize(fun = costfn, x0 = initial_theta, args = (X, y),
                method = 'TNC', jac = gradient, options = {'maxiter': 400})
    cost = res.fun
    theta = res.x

    # Print theta to screen
    print("Cost at theta found by op.minimize: %.3f" % cost)
    print("  theta:\n\t[%.3f, %.3f, %.3f]" % (theta[0], theta[1], theta[2]))

    plotDecisionBoundary(theta, X, y)
    plt.xlabel('Exam 1 score')
    plt.ylabel('Exam 2 score')
    plt.legend(['Admitted', 'Not admitted'])
    plt.show()

    input("Program paused. Press enter to continue.")

## ============== Part 4: Predict and Accuracies ==============
    prob = sigmoid(np.matmul(np.array([1, 45, 85]), theta))
    print("For a student with scores 45 and 85, we predict an admission probability of %.3f" % prob)

    # Compute accuracy on our training set
    p = predict(theta, X)

    print("Training accuracy: %.3f" % (np.mean(p == np.reshape(y, [m, 1])) * 100))
    

File: machine-learning-ex2/r/ex2.r
## Machine Learning Online Class (R implementation) - Exercise 2: Logistic Regression
require('rdetools')
rm(list = ls())  # Clear environment variables
lapply(list.files(pattern = "[.]R$", recursive = TRUE), source)  # load all fn's

data = read.csv('ex2data1.txt', header = F)
X = data[,1:2]
y = data[,3]

## ==================== Part 1: Plotting ====================
cat("Plotting data with + indicating (y=1) examples and o indicating (y=0) examples.\n")

plotData(X, y, 'Exam 1 score', 'Exam 2 score')
legend(x = 40, y = 100, c('Admitted', 'Not admitted'), pch = c(3, 20), col = c('black', 'orange'))

readline(prompt = "Program paused. Press enter to continue.")

## ============ Part 2: Compute Cost and Gradient ============
mn = dim(X)
m = mn[1]; n = mn[2]

# Add intercept term to x and X_test
X = as.matrix(cbind(rep(1, m), X))

# Initialize fitting parameters
initial_theta = rep(0, n + 1)

# Compute and display initial cost and gradient
cg = costFunction(initial_theta, X, y)
cost = cg$cost
grad = cg$grad

cat(paste0("Cost at initial theta (zeros): ", cost, "\n"))
cat("Gradient at initial theta (zeros):\n")
cat(paste0("  [", grad[1], ", ", grad[2], ", ", grad[3], "]\n"))

readline(prompt = "Program paused. Press enter to continue.")

## ============= Part 3: Optimizing using fminunc  =============
res = optim(initial_theta, fn = costfn, gr = gradient, X = X, y = y, control = list(maxit = 400))
theta = res$par
cost = res$value

cat(paste0("Cost at theta found by optim(): ", cost, "\n"))
cat("Gradient at theta:\n")
cat(paste0("  [", theta[1], ", ", theta[2], ", ", theta[3], "]\n"))

# Plot decision boundary
plotDecisionBoundary(theta, X, y, 'Exam 1 score', 'Exam 2 score')
legend(x = 40, y = 100, c('Admitted', 'Not admitted'), pch = c(3, 20), col = c('black', 'orange'))

readline(prompt = "Program paused. Press enter to continue.")

## ============== Part 4: Predict and Accuracies ==============
prob = sigmoid(c(1, 45, 85) %*% theta) * 100
cat(paste0("For a student with scores 45 and 85, we predict an admission probability of ", prob, "%.\n"))

# Compute accuracy on our training set
p = predict(theta, X)

cat(paste0("Training accuracy: ", mean(p == y) * 100, "%\n"))

